### DISCUSSION TITLE
how to use modbus error to create validations

### DISCUSSION THREAD

**Discussion by jagonzalezga** [eModbus/eModbus#219] [Q&A] [answered]
hello, my problem is that I don't know how to handle the modbus error to make validations in my favor and in this way do it through if statements, so if someone could help me it would be of great help, in short I want to use the modbus error to generate validations is all, thanks in advance.

**Comment by Miq1** [comment]
The ModbusError is returned whenever the lib has detected something being against the Modbus standard or violating internal guidelines. It always will be one of those mentioned [here](https://github.com/eModbus/eModbus/blob/master/src/ModbusError.h).

You can check the returned error code with any of the defined error codes in an if clause.

**Reply by jagonzalezga** [reply]
Now my question is if I can use outside of the handleError function, the truth is I'm new to this and I want to know if you could help me with a simple example please, thank you very much for your answer Miq1.

**Comment by Miq1** [comment, marked as answer]
The error code is a simple uint8_t wide value and as such can be stored, forwarded and whatever you like to do with it.
I assume you rather will need some other insight, though. The ``handleError`` and ``handleData`` functions used in almost all the examples are applications of the asynchronous character of the request mechanism. That means that the flow of your program is not held by a request until the response has arrived, but it will be handled by a background task. Only if a response was received (or a timeout has happened) the ``handle...`` callbacks are called.
It is your task to grab and process the returned data there for the main loop to use it eventually.
The opposite mechanism are synchronous requests, where the request is sent and the program is held while waiting for a response. This response is returned by the call directly. If you will refer to the docs you can see all asynchronous requests do have a synchronous pendant. This is less suited for most use cases, since normally the sketch will have a main task to solve that would be difficult if waiting for some Modbus responses all the time.

**Reply by jagonzalezga** [reply]
thank you for your answer it helped me a lot
